src/xdsl/irdl.py

```diff
@@ -170,28 +202,43 @@ def irdl_to_attr_constraint(irdl: Any) -> AttrConstraint:
 
     # Annotated case
     # Each argument of the Annotated type correspond to a constraint to satisfy.
-    if typing.get_origin(irdl) == Annotated:
-        constraints = []
-        for arg in typing.get_args(irdl):
+    if get_origin(irdl) == Annotated:
+        constraints: List[AttrConstraint] = []
+        for arg in get_args(irdl):
             # We should not try to convert IRDL annotations, which do not
             # correspond to constraints
             if isinstance(arg, IRDLAnnotations):
                 continue
             constraints.append(irdl_to_attr_constraint(arg))
         if len(constraints) > 1:
-            return AndConstraint(constraints)
+            return AllOf(constraints)
         return constraints[0]
 
     # Attribute class case
     # This is a coercion for an `BaseAttr`.
     if isclass(irdl) and issubclass(irdl, Attribute):
         return BaseAttr(irdl)
 
+    # yapf: disable
```

Comment:
I'm not a big fan of complexifying IRDL with this. This is because we would have the same issue with `Dict`, or any generic data structure we could put in `Data`. This would mean that any other case would have to be encoded in IRDL, which is not really extensible.

I feel that instead, we could add some fields and methods to `ArrayAttr`, and make IRDL collect this information to produce these constraints. I'm gonna try a bit tomorrow to see how it look.

Reply:
I actually thought about this as an alternative.
Reasons, why I went with this design:
  1. this fixes the type of the `data` field to be the once we want (e.g. when we write `something.data[0].data` where `something: Data[List[IntAttr]]` then we get an `int` back and the type checker knows this)
  2. there is nothing more to be said, the constraint can be inferred automatically (as the implementation shows)
  3. I deliberately made the implementation in `IRDL` not specific to `ArrayAttr` but generic to `Data[List[A]]`, so _any_ subclass of `Data` with a `List[A]`.
  Given this type, the `data` field will be `List[A]` and therefore the constraint to verify is that all elements of the list must satisfy the constraint of `A`.
  All of this is just a continuation of what `Data` does already.
  
  We could also implement a constraint that verifies the equivalent for `Dict[K, V]` checking that all keys satisfy the `K` constraint and all associated values the `V` constraint.
  
  Furthermore, we could restrict what you are allowed to use as a type-argument with `Data`, e.g. `Attribute | List | Dict` but nothing more ...

Reply:
I completely agree with points `1` and `2`, and don't intend to change them!
The only thing I don't really like is that `IRDL` now needs to be modified for any new data structure that users might use.
I'll try something now to show an example on how I thought doing this!

Reply:
I wrote a pull request that should explain what I meant: #105
Feel free to merge it to this PR if you think this makes sense, otherwise we can discuss it on that PR!

Reply:
I had a look at the PR.
You clearly succeeded at keeping `1`, but I don't think you managed to keep `2`, as you now have to implement the `generic_constraint_coercion` function in `ArrayAttr`.

I am ok with the change, as I can see your point of keeping `IRDL` free of "special" cases, even though I think that you now open the door to allowing people to misuse the API more easily (as they _can_ define a subclass of `Data[List[A]]` but forget to implement `generic_constraint_coercion`, or implement it wrongly).

Reply:
Oh, I think I misunderstood point `2`! I don't think there is actually any way to go around it, in the sense that I don't think we can handle most cases in IRDL, since each new case will require a change in IRDL (For instance, I don't think two users will have the same `ArrayData` using `List`).

We are sure that users won't forget to implement the `generic_constraint_coercion`, since IRDL won't accept any generic Data without implementing this (This is because `IRDL` cannot translate generic types to attribute constraints). However, we can't prevent them to implementing them wrongly I believe.

I'll add on the other PR some features to document it better when user forget to implement `generic_constraint_coercion`!